src/lazy-wait/main.tsx

```diff
@@ -0,0 +1,55 @@
+import React, { useContext, useEffect, useRef } from 'react';
+import type { ReactNode } from 'react';
+
+import { UntilContext } from './context';
+import type { UntilSubscriber } from './context';
+
+export type LazyWaitProps = {
+  until: boolean;
+  children: ReactNode;
+};
+
+export const LazyWait = ({ until, children }: LazyWaitProps) => {
+  const closestUntil = useContext(UntilContext);
+  const untilProp = useRef(until);
+  const value = useRef(until && closestUntil.value.current);
+  const subscribers = useRef<Set<UntilSubscriber>>(new Set());
+  const api = useRef({
+    subscribe: (subscriber: UntilSubscriber) => {
+      subscribers.current.add(subscriber);
+
+      return () => {
+        subscribers.current.delete(subscriber);
+      };
+    },
+    value,
+  });
+
+  untilProp.current = until;
+
+  useEffect(() => {
+    // Notify subscribers when until prop changes
+    value.current = closestUntil.value.current && until;
+    for (const subscriber of subscribers.current) {
+      subscriber(value.current);
+    }
+  }, [closestUntil, until]);
```

Comment:
Wonder if the separation between until prop change and parent change helps clarity or not.
As these two use effects could be written as:

```js
useEffect(() => {
  // Notify subscribers when until prop or closet until value changes
  const notify = (nextUntil) => {
    value.current = nextUntil && until;
    for (const subscriber of subscribers.current) {
      subscriber(value.current);
    }
  }
  
  const unsubscribe = closestUntil.subscribe(notify);
  notify(closestUntil.value.current);
  return unsubscribe;
}, [closestUntil, until]);
```

So we might not need to store until prop in a ref either